File: src/main/java/cz/wa/secretagent/game/controller/menucreator/GameMenuCreator.java
//package cz.wa.secretagent.game.controller.menucreator;
//
//import java.io.Serializable;
//import java.util.ArrayList;
//import java.util.List;
//
//import org.springframework.beans.factory.annotation.Required;
//
//import cz.wa.secretagent.game.starter.MapStarter;
//import cz.wa.secretagent.menu.builder.TextButtonDescriptor;
//import cz.wa.secretagent.menu.window.GFrame;
//import cz.wa.secretagent.menu.window.component.selectable.GButtonListener;
//
///**
// * Methods for creating menus in game. Caches all frames.
// *
// * @author Ondrej Milenovsky
// */
//public class GameMenuCreator extends GeneralMenuCreator implements Serializable {
//    private static final long serialVersionUID = 114350147422291071L;
//
//    private MapStarter mapStarter;
//
//    private transient GFrame quitToMenuDialog;
//    private transient GFrame levelMainMenu;
//    private transient GFrame islandMainMenu;
//
//    private void init() {
//        if (quitToMenuDialog == null) {
//            quitToMenuDialog = createQuitToMenuDialog();
//        }
//    }
//
//    public GFrame getLevelMainMenu() {
//        if (levelMainMenu == null) {
//            levelMainMenu = createLevelMainMenu();
//        }
//        return levelMainMenu;
//    }
//
//    public GFrame getIslandMainMenu() {
//        if (islandMainMenu == null) {
//            islandMainMenu = createIslandMainMenu();
//        }
//        return islandMainMenu;
//    }
//
//    private GFrame createLevelMainMenu() {
//        init();
//        final GFrame returnToIslandDialog = createReturnToIslandDialog();
//
//        // buttons
//        List<TextButtonDescriptor> buttons = new ArrayList<TextButtonDescriptor>();
//        buttons.add(createSettingsButton());
//        buttons.add(new TextButtonDescriptor("Return to island", new GButtonListener() {
//            @Override
//            public void actionPerformed() {
//                if (gameSettings.confirmDialogs) {
//                    returnToIslandDialog.setSelectedIndex(0);
//                    worldHolder.getMenuHolder().addFrame(returnToIslandDialog);
//                } else {
//                    quitToIsland();
//                }
//            }
//        }));
//        buttons.add(createQuitToMenuButton());
//        buttons.add(createExitButton());
//
//        // frame
//        GButtonListener closeListener = new GButtonListener() {
//            @Override
//            public void actionPerformed() {
//                worldHolder.getMenuHolder().removeTopFrame();
//                worldHolder.getWorld().setRunning(true);
//            }
//        };
//        GFrame dialog = dialogBuilder.createDialog("Game paused", buttons, frameColor, closeListener, false);
//
//        return dialog;
//    }
//
//    private void quitToIsland() {
//        mapStarter.startIslandMap();
//        worldHolder.getMenuHolder().removeAllFrames();
//    }
//
//    private TextButtonDescriptor createExitButton() {
//        return new TextButtonDescriptor("Exit the game", new GButtonListener() {
//            @Override
//            public void actionPerformed() {
//                if (gameSettings.confirmDialogs) {
//                    GFrame exitDialog = getExitDialog();
//                    exitDialog.setSelectedIndex(0);
//                    worldHolder.getMenuHolder().addFrame(exitDialog);
//                } else {
//                    exitGame();
//                }
//            }
//        });
//    }
//
//    private TextButtonDescriptor createQuitToMenuButton() {
//        return new TextButtonDescriptor("Quit to main menu", new GButtonListener() {
//            @Override
//            public void actionPerformed() {
//                if (gameSettings.confirmDialogs) {
//                    quitToMenuDialog.setSelectedIndex(0);
//                    worldHolder.getMenuHolder().addFrame(quitToMenuDialog);
//                } else {
//                    quitToMenu();
//                }
//            }
//        });
//    }
//
//    private void quitToMenu() {
//        worldHolder.getMenuHolder().removeAllFrames();
//        mapStarter.startMainMenu();
//    }
//
//    private TextButtonDescriptor createSettingsButton() {
//        return new TextButtonDescriptor("Settings", new GButtonListener() {
//            @Override
//            public void actionPerformed() {
//                GFrame settingsMenu = getSettingsMenu();
//                settingsMenu.setSelectedIndex(0);
//                worldHolder.getMenuHolder().addFrame(settingsMenu);
//            }
//        });
//    }
//
//    private GFrame createIslandMainMenu() {
//        init();
//        // buttons
//        List<TextButtonDescriptor> buttons = new ArrayList<TextButtonDescriptor>();
//        buttons.add(createSettingsButton());
//        buttons.add(createQuitToMenuButton());
//        buttons.add(createExitButton());
//
//        // frame
//        GButtonListener closeListener = new GButtonListener() {
//            @Override
//            public void actionPerformed() {
//                worldHolder.getMenuHolder().removeTopFrame();
//                worldHolder.getWorld().setRunning(true);
//            }
//        };
//        GFrame dialog = dialogBuilder.createDialog("Game paused", buttons, frameColor, closeListener, false);
//
//        return dialog;
//    }
//
//    private GFrame createQuitToMenuDialog() {
//        return createConfirmDialog("Quit to main menu?", new GButtonListener() {
//            @Override
//            public void actionPerformed() {
//                quitToMenu();
//            }
//        }, frameColor);
//    }
//
//    private GFrame createReturnToIslandDialog() {
//        return createConfirmDialog("Return to island map?", new GButtonListener() {
//            @Override
//            public void actionPerformed() {
//                quitToIsland();
//            }
//        }, frameColor);
//    }
//
//    public MapStarter getMapStarter() {
//        return mapStarter;
//    }
//
//    @Required
//    public void setMapStarter(MapStarter mapStarter) {
//        this.mapStarter = mapStarter;
//    }
//
//}

File: src/main/kotlin/secretAgent/game/controller/menuCreator/GameMenuCreator.kt
package secretAgent.game.controller.menuCreator

import secretAgent.game.starter.MapStarter
import secretAgent.menu.TextButtonDescriptor
import secretAgent.menu.window.GFrame
import secretAgent.menu.window.component.GButtonListener
import java.io.Serializable

/**
 * Methods for creating menus in game. Caches all frames.
 *
 * @author Ondrej Milenovsky
 */
class GameMenuCreator : GeneralMenuCreator(), Serializable {

    lateinit var mapStarter: MapStarter

    @Transient
    private lateinit var quitToMenuDialog: GFrame
    @Transient
    private lateinit var levelMainMenu: GFrame
    @Transient
    private lateinit var islandMainMenu: GFrame

    fun getLevelMainMenu(): GFrame { // TODO lazy
        if(!::levelMainMenu.isInitialized)
            levelMainMenu = createLevelMainMenu()
        return levelMainMenu
    }

    fun getIslandMainMenu(): GFrame { // TODO lazy
        if(!::islandMainMenu.isInitialized)
            islandMainMenu = createIslandMainMenu()
        return islandMainMenu
    }

    private fun init() {
        if (!::quitToMenuDialog.isInitialized)
            quitToMenuDialog = createQuitToMenuDialog()
    }

    private fun createLevelMainMenu(): GFrame {
        init()
        val returnToIslandDialog = createReturnToIslandDialog()

        // buttons
        val buttons = arrayListOf(
                createSettingsButton(),
                TextButtonDescriptor("Return to island") {
                    if (gameSettings.confirmDialogs) {
                        returnToIslandDialog.selectedIndex = 0
                        worldHolder.menuHolder.addFrame(returnToIslandDialog)
                    } else
                        quitToIsland()
                },
                createQuitToMenuButton(),
                createExitButton())

        // frame
        val closeListener: GButtonListener = {
            worldHolder.menuHolder.removeTopFrame()
            worldHolder.world.isRunning = true
        }

        return dialogBuilder.createDialog("Game paused", buttons, frameColor, closeListener, false)
    }

    private fun quitToIsland() {
        mapStarter.startIslandMap()
        worldHolder.menuHolder.removeAllFrames()
    }

    private fun createExitButton(): TextButtonDescriptor {
        return TextButtonDescriptor("Exit the game") {
            if (gameSettings.confirmDialogs) {
                val exitDialog = exitDialog
                exitDialog.selectedIndex = 0
                worldHolder.menuHolder.addFrame(exitDialog)
            } else
                exitGame()
        }
    }

    private fun createQuitToMenuButton(): TextButtonDescriptor {
        return TextButtonDescriptor("Quit to main menu") {
            if (gameSettings.confirmDialogs) {
                quitToMenuDialog.selectedIndex = 0
                worldHolder.menuHolder.addFrame(quitToMenuDialog)
            } else
                quitToMenu()
        }
    }

    private fun quitToMenu() {
        worldHolder.menuHolder.removeAllFrames()
        mapStarter.startMainMenu()
    }

    private fun createSettingsButton() = TextButtonDescriptor("Settings") {
        val settingsMenu = settingsMenu
        settingsMenu.selectedIndex = 0
        worldHolder.menuHolder.addFrame(settingsMenu)
    }

    private fun createIslandMainMenu(): GFrame {
        init()
        // buttons
        val buttons = arrayListOf(
                createSettingsButton(),
                createQuitToMenuButton(),
                createExitButton())
        // frame
        val closeListener: GButtonListener = {
            worldHolder.menuHolder.removeTopFrame()
            worldHolder.world.isRunning = true
        }

        return dialogBuilder.createDialog("Game paused", buttons, frameColor, closeListener, false)
    }

    private fun createQuitToMenuDialog(): GFrame = createConfirmDialog("Quit to main menu?", frameColor) { quitToMenu() }

    private fun createReturnToIslandDialog(): GFrame = createConfirmDialog("Return to island map?", frameColor) { quitToIsland() }

    companion object {
        private const val serialVersionUID = 114350147422291071L
    }
}